### PR TITLE
Add OOP to manual

### PR DESCRIPTION
**Pull request by brianharvey**

### CODE REVIEW

**Comment by jrincayc** [comment]
Thanks, I'll review this later this week.

**Comment by brianharvey** [comment]
This is taken from the Object Logo manual and I'm not 100% sure that the Berkeley Logo version works exactly the same way.  I don't quite understand, for example, why procedure foo can't say usual.bar, but only usual.foo; I'm guessing it's some idea of good hygeine rather than a real technical obstacle.  And I'm not sure we have the thing about having a default turtle separate from the current object, so if you enter a turtle command with a non-turtle current object, the method in the Logo object redirects the command to the default turtle.  I don't have an oop version running on my machine -- I guess I'll do that next. :)

**Comment by pahihu** [comment]
Hi,

I’ve skimmed over the docs. One feature is definitely not supported: calling usual.foo when foo is not defined in the parent.
This is ugly, because we don’t know how to parse usual.foo because the arity of the command/function is not known.

I’ve just typed in a simple example, which calls usual.bar in xyz procedure:

to bar
  pr [default bar]
end

make "foo something

ask :foo [to xyz]
  usual.bar
  pr [foo xyz]
end

make "i kindof :foo
ask :i [xyz]


Andras Pahi

**Comment by brianharvey** [comment]
@pahihu Why is that a problem?  Like everything else, you don't parse the instruction line in xyz until it is first run, and at that time you can find foo's parent (which is Logo in your example) and look up bar there to find its arity.  It's true that in a multiple-parent situation the "usual.bar" in xyz might refer to i's second parent rather than to Logo, and then you might have to reparse the line.  I would hope that in that (rare) situation you don't change the saved parsing but just do a temporary parsing.  Alternatively, I think it would be okay to require that all accessible versions of bar have the same arity when usual.bar appears.

(I'm trying to avoid reading the code because I'm drowning in non-Logo work, but I'll try to squeeze it in if I have to...)  Thanks!

**Comment by pahihu** [comment]
Well, I'm actually writing in “Hungrish” - my native language is Hungarian.

My comment was on the following line:
> "I don't quite understand, for example, why procedure foo can't say usual.bar, but only usual.foo;
> I'm guessing it's some idea of good hygeine rather than a real technical obstacle.”

What you’ve described is working as expected, my example verifies this.
Even parsing usuals with different arity should work.

From the modified usermanual.texi:
> If there is no inherited procedure of the same name, then calling
> @code{usual.___} does nothing if used as a command, or outputs an empty
> list if used as an operation.

This will not work, because we don’t know the arity of the non-existent command/operation.


Thanks,
Andras Pahi

**Comment by brianharvey** [comment]
I see.  Does that mean you can pronounce the name of that metro ticket that's good for three stops in the downtown zone within 15 minutes?  (Did I remember that correctly?)  I could never learn it; I just pointed at the brochure at the ticket desk.  :)

I guess what Object Logo does is assume the same arity as the procedure in the current object.  That must be the reason for their restriction about usual.foo allowed only in foo.

**Comment by pahihu** [comment]
Regarding the unknown arity: I think I can make a fix working as Object Logo.
Finally I dig out an old copy of Object Logo Student Edition I try to make it fly on the Mini vMac emulator.

Just a side note: we’ve made a progress ;)
The metro ticket is good even when changing metro lines within 80 minutes…

File: docs/usermanual.texi [hunk 1]
```diff
@@ -1459,8 +1450,474 @@ parentheses are separate members of the output.  Note that sublists of a
...
+
+@example
+SOMETHING
+@end example
+
+creates and outputs an object whose parent is the @code{Logo} object.
+@code{make "foo something} is equivalent to @code{make "foo kindof
+logo}.
+
+@xref{logo}.
```
Comment (jrincayc) [inline docs/usermanual.texi]: There does not seem to be a ```@anchor``` or ```@node``` to go with this reference.

File: docs/usermanual.texi [hunk 2]
```diff
@@ -1459,8 +1450,474 @@ parentheses are separate members of the output.  Note that sublists of a
...
+
+Initializes a new object.  It is run automatically by @code{oneof}.
+
+The default @code{exist} procedure creates object-local variables with
+the odd-numbered elements of @code{initlist} as names, and the
+corresponding even-numbered elements as values.  Thus
+
+@example
+make "obj (oneof logo "foo "bar "baz "garply)
+@example
```
Comment (jrincayc) [inline docs/usermanual.texi]: Should this be ```@end example```

Comment (jrincayc) [reply docs/usermanual.texi]: Fixed.

**Comment by jrincayc** [comment]
I got the texi to compile to a pdf :)  Still has a lot of warnings however.

**Comment by brianharvey** [comment]
Awesome.  Are the warnings all from the new section?

**Comment by jrincayc** [comment]
Most of them are:
```
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019) (preloaded format=pdfetex)
 restricted \write18 enabled.
entering extended mode
(./usermanual.texi (/usr/share/texmf/tex/texinfo/texinfo.tex
Loading texinfo [version 2019-02-16.14]: pdf, fonts, markup, glyphs,
page headings, tables, conditionals, indexing, sectioning, toc, environments,
defuns, macros, cross references, insertions,
(/usr/share/texlive/texmf-dist/tex/generic/epsf/epsf.tex
This is `epsf.tex' v2.7.4 <14 February 2011>
) localization, formatting, and turning on texinfo input format.) [1{/usr/share
/texlive/texmf-dist/fonts/map/pdftex/updmap/pdftex.map}]
(/home/jjc/software/ucblogo-code/docs/usermanual.toc) [-1] [-2]
(/home/jjc/software/ucblogo-code/docs/usermanual.toc [-3] [-4] [-5] [-6]
[-7] [-8] [-9] [-10]) [-11] [-12]
(/home/jjc/software/ucblogo-code/docs/usermanual.toc)
(/home/jjc/software/ucblogo-code/docs/usermanual.toc) Chapter 1
(/home/jjc/software/ucblogo-code/docs/usermanual.aux)
Writing index file usermanual.cp
[1] [2] [3] [4] [5] [6] [7] Chapter 2 [8] [9] [10] [11] [12] [13] [14] [15]
[16] [17] Chapter 3 [18] [19]pdfTeX warning (ext4): destination with the same i
dentifier (name{CONSTRUCTORS}) has been already used, duplicate ignored
<to be read again> 
                   {
@setref #1#2->@pdfmkdest {#1}@iflinks {
                                       @requireauxfile @atdummies @def @valu...

@donoderef ...empty @else @setref {@lastnode }{#1}
                                                  @global @let @lastnode =@e...

@sectionheading ...nlevel }{#1}{#4}@donoderef {#3}
                                                  @nobreak @vbox {@hyphenpen...

@genhead ...level @chapterzzz {#3}@or @seczzz {#3}
                                                  @or @numberedsubseczzz {#3...
l.1542 @section Constructors
                             l.1572: Undefined cross reference `logo-snt'.
l.1572: Undefined cross reference `logo-snt'.
l.1572: Undefined cross reference `logo-pg'.
l.1595: Undefined cross reference `exist-snt'.
l.1595: Undefined cross reference `exist-snt'.
l.1595: Undefined cross reference `exist-pg'.pdfTeX warning (ext4): destination
 with the same identifier (name{MUTATORS}) has been already used, duplicate ign
ored
<to be read again> 
                   {
@setref #1#2->@pdfmkdest {#1}@iflinks {
                                       @requireauxfile @atdummies @def @valu...

@donoderef ...empty @else @setref {@lastnode }{#1}
                                                  @global @let @lastnode =@e...

@sectionheading ...nlevel }{#1}{#4}@donoderef {#3}
                                                  @nobreak @vbox {@hyphenpen...

@genhead ...level @chapterzzz {#3}@or @seczzz {#3}
                                                  @or @numberedsubseczzz {#3...
l.1598 @section Mutators
                         [20] l.1644: Undefined cross reference `oneof-snt'.
l.1644: Undefined cross reference `oneof-snt'.
l.1644: Undefined cross reference `oneof-pg'.
l.1644: Undefined cross reference `usual-snt'.
l.1644: Undefined cross reference `usual-snt'.
l.1644: Undefined cross reference `usual-pg'.
l.1644: Undefined cross reference `havemake-snt'.
l.1644: Undefined cross reference `havemake-snt'.
l.1644: Undefined cross reference `havemake-pg'.
l.1658: Undefined cross reference `havemake-snt'.
l.1658: Undefined cross reference `havemake-snt'.
l.1658: Undefined cross reference `havemake-pg'.
l.1677: Undefined cross reference `have-snt'.
l.1677: Undefined cross reference `have-snt'.
l.1677: Undefined cross reference `have-pg'.
l.1677: Undefined cross reference `make-snt'.
l.1677: Undefined cross reference `make-snt'.
l.1677: Undefined cross reference `make-pg'.pdfTeX warning (ext4): destination 
with the same identifier (name{SELECTORS}) has been already used, duplicate ign
ored
<to be read again> 
                   {
@setref #1#2->@pdfmkdest {#1}@iflinks {
                                       @requireauxfile @atdummies @def @valu...

@donoderef ...empty @else @setref {@lastnode }{#1}
                                                  @global @let @lastnode =@e...

@sectionheading ...nlevel }{#1}{#4}@donoderef {#3}
                                                  @nobreak @vbox {@hyphenpen...

@genhead ...level @chapterzzz {#3}@or @seczzz {#3}
                                                  @or @numberedsubseczzz {#3...
l.1681 @section selectors
                          [21]
l.1754: Undefined cross reference `mynames-snt'.
l.1754: Undefined cross reference `mynames-snt'.
l.1754: Undefined cross reference `mynames-pg'.
l.1780: Undefined cross reference `myprocs-snt'.
l.1780: Undefined cross reference `myprocs-snt'.
l.1780: Undefined cross reference `myprocs-pg'. [22]
l.1806: Undefined cross reference `ask-snt'.
l.1806: Undefined cross reference `ask-snt'.
l.1806: Undefined cross reference `ask-pg'.
l.1822: Undefined cross reference `talkto-snt'.
l.1822: Undefined cross reference `talkto-snt'.
l.1822: Undefined cross reference `talkto-pg'.pdfTeX warning (ext4): destinatio
n with the same identifier (name{QUERIES}) has been already used, duplicate ign
ored
<to be read again> 
                   {
@setref #1#2->@pdfmkdest {#1}@iflinks {
                                       @requireauxfile @atdummies @def @valu...

@donoderef ...empty @else @setref {@lastnode }{#1}
                                                  @global @let @lastnode =@e...

@sectionheading ...nlevel }{#1}{#4}@donoderef {#3}
                                                  @nobreak @vbox {@hyphenpen...

@genhead ...level @chapterzzz {#3}@or @seczzz {#3}
                                                  @or @numberedsubseczzz {#3...
l.1825 @section Queries
                        [23] Chapter 4 [24] [25] [26] [27] [28] [29] [30]
[31] [32] [33] Chapter 5 [34] [35] [36] [37] [38] [39] Chapter 6 [40]
Chapter 7 [41] [42] [43] [44] [45] [46] [47] [48] [49] [50] [51] [52]
Chapter 8 [53] [54] [55] [56] l.4682: Undefined cross reference `text-snt'.
l.4682: Undefined cross reference `text-snt'.
l.4682: Undefined cross reference `text-pg'. [57] [58] [59] [60] [61] [62]
[63] [64] [65] [66] [67] [68] [69] [70] Chapter 9 [71] [72] [73] [74] [75]
[76] [77] [78] [79] [80] [81] [82] [83] [84] [85] [86] Chapter 10 [87] [88]
[89] [90] Chapter 11 [91] [92] [93] Chapter 12 [94] [95] [96] Chapter 13
[97] [98] [99] [100] (INDEX) [101] [102] [103] [104] [105] [106] [107] [108] )
(see the transcript file for additional information)pdfTeX warning (dest): name
{text} has been referenced but does not exist, replaced by a fixed one

pdfTeX warning (dest): name{talkto} has been referenced but does not exist, rep
laced by a fixed one

pdfTeX warning (dest): name{ask} has been referenced but does not exist, replac
ed by a fixed one

pdfTeX warning (dest): name{myprocs} has been referenced but does not exist, re
placed by a fixed one

pdfTeX warning (dest): name{mynames} has been referenced but does not exist, re
placed by a fixed one

pdfTeX warning (dest): name{make} has been referenced but does not exist, repla
ced by a fixed one

pdfTeX warning (dest): name{have} has been referenced but does not exist, repla
ced by a fixed one

pdfTeX warning (dest): name{havemake} has been referenced but does not exist, r
eplaced by a fixed one

pdfTeX warning (dest): name{usual} has been referenced but does not exist, repl
aced by a fixed one

pdfTeX warning (dest): name{oneof} has been referenced but does not exist, repl
aced by a fixed one

pdfTeX warning (dest): name{exist} has been referenced but does not exist, repl
aced by a fixed one

pdfTeX warning (dest): name{logo} has been referenced but does not exist, repla
ced by a fixed one

</usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmb10.pfb></usr/s
hare/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmbx12.pfb></usr/share/t
exlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmmi10.pfb></usr/share/texlive
/texmf-dist/fonts/type1/public/amsfonts/cm/cmmi12.pfb></usr/share/texlive/texmf
-dist/fonts/type1/public/amsfonts/cm/cmmi9.pfb></usr/share/texlive/texmf-dist/f
onts/type1/public/amsfonts/cm/cmr10.pfb></usr/share/texlive/texmf-dist/fonts/ty
pe1/public/amsfonts/cm/cmr12.pfb></usr/share/texlive/texmf-dist/fonts/type1/pub
lic/amsfonts/cm/cmr7.pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsf
onts/cm/cmr9.pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/
cmsl10.pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmsl9.
pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmsltt10.pfb>
</usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmsy10.pfb></usr/
share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmti10.pfb></usr/share/
texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cmtt10.pfb></usr/share/texliv
e/texmf-dist/fonts/type1/public/amsfonts/cm/cmtt9.pfb>
Output written on usermanual.pdf (121 pages, 515310 bytes).
Transcript written on usermanual.log.
```

File: docs/usermanual.texi (hunk 1 above)
Comment (jrincayc) [reply docs/usermanual.texi]: @brianharvey I am not sure where this ref is supposed to go.  I can add an ```@anchor``` if you tell me where this should reference.  The rest of the reference problems I can fix by capitalizing them.

**Comment by jrincayc** [comment]
I fixed all the warning but the one from reference to logo for building the pdf.  I still have to fix the makeinfo errors.

**Comment by brianharvey** [comment]
I think it's a missing primitive rather than an error in the manual.  The procedure "logo" takes no inputs and reports the root object of the tree -- all objects are descendents of the Logo object.  So, equivalently, the "something" primitive is equivalent to "kindof logo".
When Things Slow Down™ we should go through the graph of systemy objects (turtle, window, menu, input stream, etc.) and decide which ones to implement.

**Comment by jrincayc** [comment]
> I think it's a missing primitive rather than an error in the manual. The procedure "logo" takes no inputs and reports the root object of the tree -- all objects are descendents of the Logo object. So, equivalently, the "something" primitive is equivalent to "kindof logo".
> When Things Slow Down™ we should go through the graph of systemy objects (turtle, window, menu, input stream, etc.) and decide which ones to implement.

Okay, I commented out the reference to logo and added a comment stating that logo the primitive  should be documented.

I also fixed all the makeinfo errors.

**Comment by jrincayc** [comment]
@brianharvey Thank you for creating the documentation.

**Comment by brianharvey** [comment]
No need to thank me!  I'm delighted that Berkeley Logo is being developed again.  This was the least I could do.

**Pull request merged**